builtin/consul/config_sourcer.go

```diff
@@ -0,0 +1,410 @@
+	for _, req := range reqs {
+		result := &pb.ConfigSource_Value{Name: req.Name}
+		results = append(results, result)
+
+		// Decode our configuration
+		kvReq := reqConfig{
+			// default to allowing stale queries
+			AllowStale: true,
+		}
+		if err := mapstructure.WeakDecode(req.Config, &kvReq); err != nil {
+			result.Result = &pb.ConfigSource_Value_Error{
+				Error: status.New(codes.Aborted, err.Error()).Proto(),
+			}
+
+			continue
+		}
+
+		opts := &api.QueryOptions{
+			Namespace:  kvReq.Namespace,
+			Partition:  kvReq.Partition,
+			Datacenter: kvReq.Datacenter,
+			AllowStale: kvReq.AllowStale,
+		}
+		reqLogger := log.With("key", kvReq.Key, "stale", kvReq.AllowStale)
+		if kvReq.Namespace != "" {
+			reqLogger = reqLogger.With("namespace", kvReq.Namespace)
+		}
+		if kvReq.Partition != "" {
+			reqLogger = reqLogger.With("partition", kvReq.Partition)
+		}
+		if kvReq.Datacenter != "" {
+			reqLogger = reqLogger.With("datacenter", kvReq.Partition)
+		}
+
+		cacheVal, ok := cs.cache[kvReq.cacheKey()]
+		if !ok {
+			reqLogger.Trace("querying Consul KV")
+			kvpair, meta, err := cs.client.Get(kvReq.Key, opts.WithContext(ctx))
+			if err != nil {
+				result.Result = &pb.ConfigSource_Value_Error{
+					Error: status.New(codes.Aborted, err.Error()).Proto(),
+				}
+
+				continue
+			}
+
+			refreshCtx, cancel := context.WithCancel(context.Background())
+			cacheVal = &cachedVal{
+				kvPair: kvpair,
+				cancel: cancel,
+				err:    nil,
+			}
+			cs.cache[kvReq.cacheKey()] = cacheVal
+			cs.startBlockingQuery(refreshCtx, log, kvReq, meta.LastIndex, opts)
+		}
+
+		if cacheVal.err != nil {
+			result.Result = &pb.ConfigSource_Value_Error{
+				Error: status.New(codes.Aborted, cacheVal.err.Error()).Proto(),
+			}
+		}
+
+		if cacheVal.kvPair == nil {
+			result.Result = &pb.ConfigSource_Value_Error{
+				Error: status.New(codes.NotFound, fmt.Sprintf("Configuration for key %s doesn't exist", kvReq.Key)).Proto(),
+			}
+
+			continue
+		}
+
+		result.Result = &pb.ConfigSource_Value_Value{
+			Value: string(cacheVal.kvPair.Value),
+		}
+	}
+
+	return results, nil
+}
+
+func (cs *ConfigSourcer) startBlockingQuery(ctx context.Context, logger hclog.Logger, kvReq reqConfig, lastIndex uint64, opts *api.QueryOptions) {
+	go func() {
+		opts := opts.WithContext(ctx)
+		failures := 0
+		// Ideally we would use the github.com/hashicorp/consul/api/watch package. However that package doesn't support
+		// namespaces and partitions except with the global client defaults and thus isn't suitable for this usage.
+
+		for {
+			logger.Trace("Issuing blocking query", "wait-index", lastIndex)
+			// set the wait index to use for the query
+			opts.WaitIndex = lastIndex
+			pair, meta, err := cs.client.Get(kvReq.Key, opts)
+
+			// check if we are being stopped
+			select {
+			case <-ctx.Done():
+				return
+			default:
+			}
+
+			// KV entry not updated - do nothing
+			if meta != nil && meta.LastIndex == lastIndex {
+				logger.Trace("KV value unchanged")
```

Comment:
should we sleep here for `retryInterval` duration?

Reply:
The select block at the link below, which is at the end of the for loop, should take care of waiting for us:

https://github.com/paladin-devops/waypoint/blob/5b671f4ee5809a42f2a3a68ed0aafeeccacbe8db/builtin/consul/config_sourcer.go#L333-L337

Reply:
The `continue` statement immediately below this `logger.Trace()` call shoots us back up to the top of the `for` loop.

Reply:
[Updated this to sleep for 5 seconds](https://github.com/hashicorp/waypoint/pull/4045/commits/22a277f59dd452fc5946c331babb08bcaea4f2cd) before hitting the API again!